processor/metricstarttimeprocessor/internal/truereset/adjuster.go

```diff
@@ -0,0 +1,240 @@
+// the initial points.
+type Adjuster struct {
+	jobsMap *JobsMap
+	set     component.TelemetrySettings
+}
+
+// NewAdjuster returns a new Adjuster which adjust metrics' start times based on the initial received points.
+func NewAdjuster(set component.TelemetrySettings, gcInterval time.Duration) *Adjuster {
+	return &Adjuster{
+		jobsMap: NewJobsMap(gcInterval),
+		set:     set,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their start times based on the initial and
+// previous points in the timeseriesMap.
+func (a *Adjuster) AdjustMetrics(_ context.Context, metrics pmetric.Metrics) (pmetric.Metrics, error) {
+	for i := 0; i < metrics.ResourceMetrics().Len(); i++ {
+		rm := metrics.ResourceMetrics().At(i)
+		// TODO: Produce a hash of all resource attributes, rather than just job + instance.
+		job, _ := rm.Resource().Attributes().Get(semconv.AttributeServiceName)
+		instance, _ := rm.Resource().Attributes().Get(semconv.AttributeServiceInstanceID)
+		tsm := a.jobsMap.get(job.Str(), instance.Str())
+
+		// The lock on the relevant timeseriesMap is held throughout the adjustment process to ensure that
+		// nothing else can modify the data used for adjustment.
+		tsm.Lock()
+		for j := 0; j < rm.ScopeMetrics().Len(); j++ {
+			ilm := rm.ScopeMetrics().At(j)
+			for k := 0; k < ilm.Metrics().Len(); k++ {
+				metric := ilm.Metrics().At(k)
+				switch dataType := metric.Type(); dataType {
+				case pmetric.MetricTypeGauge:
+					// gauges don't need to be adjusted so no additional processing is necessary
+
+				case pmetric.MetricTypeHistogram:
+					a.adjustMetricHistogram(tsm, metric)
+
+				case pmetric.MetricTypeSummary:
+					a.adjustMetricSummary(tsm, metric)
+
+				case pmetric.MetricTypeSum:
+					a.adjustMetricSum(tsm, metric)
+
+				case pmetric.MetricTypeExponentialHistogram:
+					a.adjustMetricExponentialHistogram(tsm, metric)
+
+				case pmetric.MetricTypeEmpty:
+					fallthrough
+
+				default:
+					// this shouldn't happen
+					a.set.Logger.Info("Adjust - skipping unexpected point", zap.String("type", dataType.String()))
+				}
+			}
+		}
+		tsm.Unlock()
+	}
+	return metrics, nil
+}
+
+func (a *Adjuster) adjustMetricHistogram(tsm *timeseriesMap, current pmetric.Metric) {
+	histogram := current.Histogram()
+	if histogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+		// Only dealing with CumulativeDistributions.
+		return
+	}
+
+	currentPoints := histogram.DataPoints()
+	for i := 0; i < currentPoints.Len(); i++ {
+		currentDist := currentPoints.At(i)
+
+		tsi, found := tsm.get(current, currentDist.Attributes())
+		if !found {
+			// initialize everything.
+			tsi.histogram.startTime = currentDist.StartTimestamp()
+			tsi.histogram.previousCount = currentDist.Count()
+			tsi.histogram.previousSum = currentDist.Sum()
+			continue
+		}
+
+		if currentDist.Flags().NoRecordedValue() {
+			// TODO: Investigate why this does not reset.
```

Comment:
[nit] Is there an issue for this already? Not sure if this is being tracked anywhere yet (I also see the TODO int he prometheusreceiver, just don't want to lose this when we remove the adjuster from there)

Reply:
I assume this is copied from existing Prometheus code?

Reply:
This is copied from the prometheus receiver.  There isn't an issue for this.